profit_analysis/analysis.py

```diff
@@ -0,0 +1,222 @@
+                .convert_dtypes(),
+                direction="nearest",
+                on=TIMESTAMP_KEY,
+            )
+
+            if len(debt_tokens_prices) > 0:
+                debt_tokens_prices[TIMESTAMP_KEY] = pd.to_datetime(
+                    debt_tokens_prices[TIMESTAMP_KEY]
+                )
+                # merge debt token prices
+                profit_with_price_token = pd.merge_asof(
+                    profit_with_price_token.astype({TIMESTAMP_KEY: PD_DATETIME_FORMAT})
+                    .sort_values(TIMESTAMP_KEY)
+                    .convert_dtypes(),
+                    debt_tokens_prices[[TIMESTAMP_KEY, PRICE_DEBT_KEY]]
+                    .astype({TIMESTAMP_KEY: PD_DATETIME_FORMAT})
+                    .sort_values(TIMESTAMP_KEY)
+                    .convert_dtypes(),
+                    direction="nearest",
+                    on=TIMESTAMP_KEY,
+                    by=TOKEN_DEBT_KEY,
+                )
+            else:
+                profit_with_price_tokens[PRICE_DEBT_KEY] = 0
+
+            profit_with_price_tokens = pd.concat(
+                [profit_with_price_tokens, profit_with_price_token]
+            )
+        except Exception as e:
+            # @TODO: save into list to add later
+            print("    Failed for token=", token)
+            print(e)
+            failures[token] = e
+    print("Finished processing all tokens")
+    profit_with_price_tokens[PRICE_DEBT_KEY] = profit_with_price_tokens[
+        PRICE_DEBT_KEY
+    ].fillna(value=0)
+    profit_with_price_tokens[AMOUNT_DEBT_KEY] = profit_with_price_tokens[
+        AMOUNT_DEBT_KEY
+    ].fillna(value=0)
+    profit_with_price_tokens["profit_usd"] = (
+        profit_with_price_tokens[AMOUNT_RECEIVED_KEY]
+        * profit_with_price_tokens[PRICE_RECEIVED_KEY]
+        - profit_with_price_tokens[AMOUNT_DEBT_KEY]
+        * profit_with_price_tokens[PRICE_DEBT_KEY]
+    )
+    profit_with_price_tokens = profit_with_price_tokens.reset_index(drop=True)
+    profit_with_price_tokens["date"] = profit_with_price_tokens[
+        TIMESTAMP_KEY
+    ].dt.normalize()
+    if save_to_csv:
+        profit_by_block.to_csv(DATA_PATH + "usd_profit.csv", index=False)
+        pd.DataFrame(failures.items(), columns=["token", "error"]).to_csv(
+            DATA_PATH + "analyze_profit_failures.csv", index=False
+        )
+    return profit_with_price_tokens
+
+
+def get_profit_by(profit_with_price_tokens, col):
+    profit_by_block = (
+        profit_with_price_tokens.groupby([col])
+        .agg({"profit_usd": ["sum", "mean", "median", "count"]})
+        .reset_index()
+    )
+    profit_by_block.columns = profit_by_block.columns.droplevel(0)
+    profit_by_block.rename(columns={"": col}, inplace=True)
+    profit_by_block.to_csv(DATA_PATH + "profit_by_" + col + ".csv", index=False)
+    return profit_by_block
+
+
+def create_web3():
+    w3_provider = web3.Web3(web3.Web3.HTTPProvider("https://polygon-rpc.com"))
```

Comment:
How about getting the RPC url as an env. variable ? We never know what might happen.

Reply:
:heavy_check_mark: